src/components/CharacterSheet.js

```diff
@@ -1,38 +1,77 @@
-import React from 'react';
+import React, { useState, useEffect } from 'react';
+import firebase from 'firebase';
 import PropTypes from 'prop-types';
 import styled from 'styled-components';
+import { Fab } from '@material-ui/core';
 import Profile from './Profile';
 import Attributes from './Attributes';
 import Skills from './Skills';
+import { TopAnchor } from './CustomStyled';
+import dungeonService from '../services/dungeonService';
 
 
-const CharacterSheet = ({ character }) => (
-  <SheetContainer>
-    <ProfileArea>
-      <Profile character={character} />
-    </ProfileArea>
-    <StatsArea>
-      <Attributes character={character} />
-    </StatsArea>
-    <SkillsArea>
-      <Skills character={character} />
-    </SkillsArea>
-    <WeaponsArea />
-    <EquipmentArea />
-  </SheetContainer>
-);
+const CharacterSheet = ({ data }) => {
+  const [character, updateCharacter] = useState(data);
+  const [isDirty, setDirty] = useState(false);
+  const [authorized, setAuthorized] = useState(false);
+
+  function update(charUpdates) {
+    setDirty(true);
+    updateCharacter(charUpdates);
+  }
+
+  async function save() {
+    setDirty(false);
+    const success = await dungeonService.saveCharacter(character);
+    if (!success) setDirty(true);
+  }
+
+  async function checkAuthorized(user) {
+    const result = await dungeonService.checkUserAuth(character.id, user.email);
+    setAuthorized(result.authorized);
+  }
+
+  useEffect(() => {
+    checkAuthorized(firebase.auth().currentUser);
+    updateCharacter(data);
+  }, [data]);
+
+  return (
+    <SheetContainer>
+      { isDirty
+        && <TopAnchor>
+          <Fab color="secondary" size="small" onClick={save}>
+            <i className="material-icons">done</i>
+          </Fab>
+           </TopAnchor>
+      }
+      <ProfileArea>
+        <Profile character={character} update={update} disabled={!authorized} />
+      </ProfileArea>
+      <StatsArea>
+        <Attributes character={character} update={update} disabled={!authorized} />
+      </StatsArea>
+      <SkillsArea>
+        <Skills character={character} />
+      </SkillsArea>
+      <WeaponsArea />
+      <EquipmentArea />
+    </SheetContainer>
+  );
+};
 
 export default CharacterSheet;
 
 CharacterSheet.propTypes = {
-  character: PropTypes.shape({
+  data: PropTypes.shape({
```

Comment:
This seems more ambiguous

Reply:
updated to `characterData`. I just want to use the `character` variable in the state so i don't want a state variable and a prop with the same name.